subprojects/core/src/main/java/org/gradle/internal/buildevents/BuildExceptionReporter.java

```diff
@@ -205,11 +211,50 @@ protected void visitLocation(String location) {
 
         @Override
         public void node(Throwable node) {
-            String message = getMessage(node);
-            if (null == node.getCause() || (StringUtils.isNotBlank(message) && !message.endsWith(NO_ERROR_MESSAGE_INDICATOR))) {
-                LinePrefixingStyledTextOutput output = getLinePrefixingStyledTextOutput(failureDetails);
-                renderStyledError(node, output);
+            if (shouldBePrinted(node)) {
+                printedNodes.add(node);
+                if (null == node.getCause() || isUsefulMessage(getMessage(node))) {
+                    LinePrefixingStyledTextOutput output = getLinePrefixingStyledTextOutput(failureDetails);
+                    renderStyledError(node, output);
+                }
+            }
+        }
+
+        /**
+         * Determines if the given node should be printed.
+         *
+         * A node should be printed iff it is not in the {@link #printedNodes} set, and it is not a
+         * transitive cause of a node that is in the set.  Direct causes will be checked, as well
+         * as each branch of {@link ContextAwareException#getReportableCauses()}s for nodes of that type.
+         *
+         * @param node the node to check
+         * @return {@code true} if the node should be printed; {@code false} otherwise
+         */
+        private boolean shouldBePrinted(Throwable node) {
+            Queue<Throwable> next = new java.util.LinkedList<>();
+            next.add(node);
+
+            while (!next.isEmpty()) {
+                Throwable curr = next.poll();
+                if (printedNodes.isEmpty()) {
```

Comment:
🤔 Couldn't this check be done once outside of the loop?

Reply:
It sure could.